docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+        :> Get '[Servant.JSON] (ConversationV version),
+[...]
+
+type family ConversationV (version :: Symbol) :: * where
+  ConversationV "v1" = Conversation
+  ConversationV "v2" = Conversation
+  ConversationV "v4" = ConversationV4
+```
+
+Note that before version `"v4"`, nothing changed for this type, so
+there was no need to introduce a new concrete data type.
+
+If the last change of a data type is entirely phased out, the type
+family turns constant and can be removed again.  If you see this in
+your code:
+
+```haskell
+type family ConversationV (version :: Symbol) :: * where
+  ConversationV "v4" = ConversationV4
+  ConversationV "v5" = ConversationV4
+  ConversationV "v6" = ConversationV4
+  ConversationV "v7" = ConversationV4
+```
+
+You can remove 'ConversationV', rename `ConversationV4` to Conversation,
+and use it in the routing table instead of `ConversationV` again, as
+before `"v4"`.
+
+
+## Changing structure
+
+Without loss of generality, we only consider additions and deletions
+of routes in this section: if you want to change the path or verb of
+an end-point, add a new path instead, and phase the old one out (now
+or in some future version).
+
+When end-points are present in some supported versions, but not in
+others, their record fields in the servant routing type needs to be
+present for all versions, but in some versions should behave as if it
+weren't.
+
+This is best solved by a new type alias:
+
+```haskell
+type NotInThisVersion = Verb 'NOTINTHISVERSION '[] NoContent
+```
+
+Then we can write a type family that can crawl a `ServantAPI` (not the
+record one, the one with `:<|>`) and drop all the routes marked as not
+existing.
+
+This will save us the trouble of writing lots of instances for
+`NotInThisVersion` (server, swagger, client, ...), and yield exactly
+the desired result:
+
+```haskell
+type ServantAPI =
+  DropNotInThisVersion
+    ( ToServantApi (Api "v1")
+        :<|> ToServantApi (Api "v2")
+        :<|> ToServantApi (Api "v4") -- v3 is broken
+    )
+```
+
+
+## Adoption of versioned APIs
+
+When API versions are introduced to a code base that has a routing
+table without versions, the question arises what to do with old
+clients or servers talking to new ones.
+
+We define a middleware that
+
+ (1) maps requests without version prefix in the path to ones that
+     have version `"v0"`.
+
+ (2) responds with a specific type of error if an unsupported version
+     is requested (so the client can re-negotiate a new version to
+     speak after an upgrade, see below).
+
+
+## Version handshake
+
+Client and server need to agree on a version to use.  The server
+provides two (kinds of) end-points for that.
+
+```
+GET /api-versions
+=> { "supported": [1, 2, 45, 119] }
+
+GET /v*/api-docs
+=> <swagger for this particular version>
+```
+
+The client developer can pull the swagger docs of a new version and
+diff it against the one they already support, and work their way
+through the changes (see below).
```

Comment:
You mean: And the changes automatically generate a new stub client

Reply:
I think so, yes.